File: 621E.py
import sys
from collections import Counter


MOD = int(10**9+7)


def main():
    class Mat:
        def __init__(self, x: int = 0, y: int = 0, mat: "Mat" = None):
            if mat is None:
                assert x and y
                self.x = x
                self.y = y
                self.mat = [[0] * y for _ in range(x)]
            else:
                self.x = mat.x
                self.y = mat.y
                self.mat = [list(i) for i in mat.mat]

        def __mul__(self, other: "Mat") -> "Mat":
            assert self.y == other.x
            ret = Mat(self.x, other.y)
            for i in range(ret.x):
                for j in range(ret.y):
                    for k in range(self.y):
                        ret.mat[i][j] += self.mat[i][k] * other.mat[k][j]
                        ret.mat[i][j] %= MOD
            return ret

        def identity(self):
            assert self.x == self.y
            for i in range(self.x):
                for j in range(self.y):
                    self.mat[i][j] = 1 if i == j else 0

        def __pow__(self, p: int) -> "Mat":
            assert self.y == self.x
            bas = Mat(mat=self)
            ret = Mat(self.x, self.y)
            ret.identity()
            while p:
                if p & 1:
                    ret = ret * bas
                bas = bas * bas
                p //= 2
            return ret

    _, b, k, x = map(int, input().split())
    frq = Counter(map(int, input().split()))
    t = Mat(x, x)
    for i in range(x):
        for j, v in frq.items():
            t.mat[i][(i * 10 + j) % x] += v

    i = Mat(1, x)
    i.mat[0][0] = 1

    print((i*t**b).mat[0][k])


with open("_input.txt", "r") as fin, open("_ok.txt", "w") as fout:
    sys.stdin = fin
    sys.stdout = fout
    main()

File: 621E.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long int ll;
typedef unsigned long long int ull;

const int N = 10, M = 65, L = 4, MOD = 1e9 + 7, HV = 151, INF = 0x3f3f3f3f;

struct Mat
{
    ull m;
    int x, y;
    vector<vector<ull>> mat;
    Mat(int _x, int _y, ull _m = MOD)
    {
        x = _x;
        y = _y;
        m = _m;
        mat = vector<vector<ull>>(x, vector<ull>(y, 0));
    }

    Mat(int _x, int _y, vector<ull> vec, ull _m = MOD)
    {
        assert(_x * _y == (int)vec.size());
        x = _x;
        y = _y;
        m = _m;
        mat = vector<vector<ull>>(x, vector<ull>(y, 0));
        for (int i = 0; i < x; i++)
        {
            for (int j = 0; j < y; j++)
            {
                mat[i][j] = vec[i * x + j] % m;
            }
        }
    }
    Mat operator*(const Mat &other) const
    {
        assert(y == other.x);
        assert(m == other.m);
        Mat ret(x, other.y, m);
        for (int i = 0; i < ret.x; i++)
        {
            for (int j = 0; j < ret.y; j++)
            {
                for (int k = 0; k < y; k++)
                {
                    // assert(ret.mat[i][j] + mat[i][k] * other.mat[k][j] < m);
                    ret.mat[i][j] = (ret.mat[i][j] + mat[i][k] * other.mat[k][j] % m) % m;
                }
            }
        }
        return ret;
    }
    void identity()
    {
        assert(x == y);
        for (int i = 0; i < x; i++)
        {
            for (int j = 0; j < y; j++)
            {
                mat[i][j] = i == j;
            }
        }
    }
    void prn()
    {
        for (int i = 0; i < x; i++)
        {
            for (int j = 0; j < y; j++)
            {
                cout << mat[i][j] << " ";
            }
            cout << endl;
        }
    }

    ll sum()
    {
        ll ret = 0;
        for (auto &&i : mat)
        {
            for (auto &&j : i)
            {
                ret = (ret + j) % m;
            }
        }
        return ret;
    }
};

Mat mat_pow(Mat b, int p)
{
    assert(b.x == b.y);
    Mat ret(b.x, b.y, b.m);
    ret.identity();
    while (p)
    {
        if (p & 1)
        {
            ret = ret * b;
        }
        b = b * b;
        p /= 2;
    }
    return ret;
}

int frq[N];

void run()
{
    int n, b, k, x;
    scanf("%d%d%d%d", &n, &b, &k, &x);
    for (int i = 0, v; i < n; i++)
    {
        scanf("%d", &v);
        frq[v]++;
    }
    vector<ull> tem(x * x, 0);
    for (int i = 0; i < x; i++)
    {
        for (int j = 0; j < N; j++)
        {
            int l = (i * 10 + j) % x;
            tem[i * x + l] += frq[j];
        }
    }
    Mat t(x, x, tem, MOD);
    tem.clear();
    tem.resize(x);
    for (int i = 0; i < x; i++)
    {
        tem[i] = i == 0;
    }
    Mat i(1, x, tem, MOD);
    printf("%llu\n", (i * mat_pow(t, b)).mat[0][k]);
}

int main()
{
    freopen("_input.txt", "r", stdin);
    freopen("_output.txt", "w", stdout);
    int t = 1;
    // scanf("%d", &t);
    while (t--)
    {
        run();
    }
    return 0;
}